plugins/python/redisgears_python.c

```diff
@@ -1746,6 +1747,572 @@ static PyObject* atomicExit(PyObject *self, PyObject *args){
+    if (!modelName) {
+        PyErr_SetString(GearsError, "Model key name was not given");
+        return NULL;
+    }
+    if(!PyUnicode_Check(modelName)){
+        PyErr_SetString(GearsError, "Model name argument must be a string");
+        return NULL;
+    }
+    RAI_Error *err;
+    RedisAI_InitError(&err);
+    const char* modelNameStr = PyUnicode_AsUTF8AndSize(modelName, NULL);
+
+    // Try to bring model from keyspace (raise an exception if it does not exist)
+    RAI_Model *model = _getModelFromKeyspace(modelNameStr, err);
+    if (model == NULL) {
+        RedisAI_FreeError(err);
+        return NULL;
+    }
+
+    // Create a modelRun op with its inputs and output keys and insert it to the DAG
+    RAI_DAGRunOp *modelRunOp = RedisAI_DAGCreateModelRunOp(model);
+    PyObject* modelInputs = GearsPyDict_GetItemString(kargs, "inputs");
+    if (!modelInputs) {
+        PyErr_SetString(GearsError, "Must specify model inputs");
+        goto cleanup;
+    }
+    PyObject* inputsIter = PyObject_GetIter(modelInputs);
+    if (!inputsIter) {
+        PyErr_SetString(GearsError, "Model inputs must be iterable");
+        goto cleanup;
+    }
+    PyObject* input;
+    while((input = (PyObject*)PyIter_Next(inputsIter)) != NULL) {
+        if(!PyUnicode_Check(input)){
+            PyErr_SetString(GearsError, "Input name must be a string");
+            goto cleanup;
+        }
+        const char* inputStr = PyUnicode_AsUTF8AndSize(input, NULL);
+        RedisAI_DAGRunOpAddInput(modelRunOp, inputStr);
+    }
+
+    PyObject* modelOutputs = GearsPyDict_GetItemString(kargs, "outputs");
+    if (!modelOutputs) {
+        PyErr_SetString(GearsError, "Must specify model outputs");
+        goto cleanup;
+    }
+    PyObject* outputsIter = PyObject_GetIter(modelOutputs);
+    if (!outputsIter) {
+        PyErr_SetString(GearsError, "Model outputs must be iterable");
+        goto cleanup;
+    }
+    PyObject* output;
+    while((output = (PyObject*)PyIter_Next(outputsIter)) != NULL) {
+        if(!PyUnicode_Check(output)){
+            PyErr_SetString(GearsError, "Output name must be a string");
+            goto cleanup;
+        }
+        const char* outputStr = PyUnicode_AsUTF8AndSize(output, NULL);
+        RedisAI_DAGRunOpAddOutput(modelRunOp, outputStr);
+    }
+    if (RedisAI_DAGAddRunOp(pyDag->dag, modelRunOp, err) != REDISMODULE_OK) {
+        PyErr_SetString(GearsError, RedisAI_GetError(err));
+        goto cleanup;
+    }
+    RedisAI_FreeError(err);
+    Py_INCREF(self);
+    return self;
+
+    cleanup:
+    RedisAI_FreeError(err);
+    RedisAI_DAGRunOpFree(modelRunOp);
+    return NULL;
+}
+
+static RAI_Script *_getScriptFromKeyspace(const char *scriptNameStr, RAI_Error *err) {
+    RedisModuleCtx *ctx = RedisModule_GetThreadSafeContext(NULL);
+    RedisGears_LockHanlderAcquire(ctx);
+    RedisModuleString* keyRedisStr = RedisModule_CreateString(ctx, scriptNameStr, strlen(scriptNameStr));
+    RedisModuleKey *key;
```

Comment:
Close the key?

Reply:
RedisAI_GetScriptFromKeyspace closes it.

Reply:
@alonre24  same question as in the model, why do you need the key param?

Reply:
Same as for model, key is now created inside the function